V2 dev

mwbenowitz opened the pull request:
This implements a set of `v2` endpoints that provide access to the new ElasticSearch index, which contains updated documents and which is updated with new data.

The `v2` endpoints also implement a more flexible structure that makes it easier to add new endpoints and to extend existing functionality. Each set of endpoints re-uses the same response and error handlers, implementing logic around each individual type of request that is received.

This also implements a basic test suite, including both unit and simple integration tests to ensure that the API is both functioning and able to return results from an ElasticSearch instance.

mwbenowitz committed:
SFR-266 Add API versioning

This introduces the ability to version the API endpoints for use with different
versions of the the ElasticSearch index and front end. Each version is configured
as its own set of roots and has the ability to point to different ElasticSearch
indexes. By defualt the API will load the v1 (earliest) API, though this default
behavior can easily be changed.

In this commit the v2 API is a simple demo endpoint that does not implement any
functional endpoints. If this approach is approved of, new endpoints will be
implemented there.

The one possible drawback to this is integration with the API Gateway. In the dev
environment, the main Gateway API includes a version number, meaning that the
endpoints potentially will have multiple version statements, which could be
confusing. However, it is unclear if this will be the case in production, or if
it is necessary to include this.

mwbenowitz committed:
Merge branch 'query-string-refactor' into v2-dev

mwbenowitz committed:
SFR-291 Implement search for new ElasticSearch Index

This implements a set of v2 search endpoints that utilize the new
ElasticSearch index containing updated records in the new data
model. Two endpoints are currently implemented, which both serve
GET and POST requests.

First, the v2/sfr/search endpoint supports basic/simple search,
and is designed to enable the current search options in the SFR
front end application. Queries are made by supplying a `field` and
`query`. Current fields supported are `title`, `author`, `subject`
and `keyword`, which is the default.

Second, the v2/sfr/work endpoint returns single work records
identified by the provieded identifier. This can be a UUID or other
identifier. However, if an identifier refers to more than one work
record an error will be raised. This endpoint is used to retrieve
single records for detail pages

mwbenowitz committed:
SFR-291 Remove unecessary console.logs from v1

mwbenowitz committed:
SFR-291 Linting Fixes

mwbenowitz committed:
SFR-291 Modify v2 handler to improve testability

mwbenowitz committed:
SFR-291 v2 Search Refactor for testability

mwbenowitz committed:
SFR-291 Work lookup refactor to enable testing

mwbenowitz committed:
SFR-291 Minor refactor to search to better handle errors

mwbenowitz committed:
SFR-291 Add Unit Test Suite

mwbenowitz committed:
SFR-291 Different identifiers for v1 and v2 endpoints

mwbenowitz committed:
SFR-291 Add basic ElasticSearch integrations

mwbenowitz committed:
Merge branch 'development' into v2-dev

gkallenberg commented on lib/client.js:
```diff
@@ -2,33 +2,33 @@ var Elastic = require('elasticsearch')
-        host: process.env.ELASTICSEARCH_HOST,
-        log: 'trace'
-    })
+  var client = new Elastic.Client({
```
`const` here?

gkallenberg commented on lib/client.js:
```diff
@@ -2,33 +2,33 @@ var Elastic = require('elasticsearch')
+    }).then(function (body) {
+      var hits = body.hits.hits
+      console.log(hits)
+    }, function (error) {
```
perhaps this should be placed in a `.catch` block after `.then`?

gkallenberg commented on routes/v1/search.js:
```diff
@@ -1,19 +1,18 @@
-
   const respond = (res, _resp, params) => {
-      var contentType = 'application/json'
+    var contentType = 'application/json'
```
`const` here

gkallenberg commented on routes/v1/search.js:
```diff
@@ -1,19 +1,18 @@
 
-      var resp = _resp
-      if (contentType !== 'text/plain') resp = JSON.stringify(_resp, null, 2)
+    var resp = _resp
```
`const` here

gkallenberg commented on routes/v1/work.js:
```diff
@@ -1,27 +1,26 @@
-
   const respond = (res, _resp, params) => {
-      var contentType = 'application/json'
+    var contentType = 'application/json'
```
`const` here

gkallenberg commented on routes/v1/work.js:
```diff
@@ -1,27 +1,26 @@
-      console.log(respLen)
-      if (respLen < 1) return handleError(res, new ElasticSearchError('Could not locate a record with that identifier'))
-      else if (respLen > 1) return handleError(res, new ElasticSearchError('Returned multiple records, identifier lacks specificity'))
+    let respLen = resp['hits']['hits'].length
```
Are the `let` params necessary since they don't appear to be reassigned within the same instance? -- same for `recordID`, `body` and `params` below. might be better to use `const` instead.

`statusCode` below should be a `let`.

mwbenowitz committed:
SFR-291 Updated Swagger docs and README

gkallenberg commented on routes/v1/search.js:
```diff
@@ -78,22 +77,22 @@ module.exports = function (app) {
-        switch(filter['field']){
-          case "year":
-            body.filter("range", "instances.pub_date", filter['value'])
+      params['filters'].map((filter) => {
```
Need to test whether `filters` is actually an array before calling `map`

gkallenberg commented on routes/v1/search.js:
```diff
@@ -44,31 +43,31 @@ module.exports = function (app) {
     // Query block, generally this will be where the main query against ES will be
     if ('queries' in params) {
-      params["queries"].map((term) => {
+      params['queries'].map((term) => {
```
Need to test whether `queries` is actually an array before calling `map`

gkallenberg commented on routes/v1/search.js:
```diff
@@ -78,22 +77,22 @@ module.exports = function (app) {
     if ('sort' in params) {
-      params["sort"].map((sort) => {
-        switch(sort["field"]){
+      params['sort'].map((sort) => {
```
Need to test whether `sort` is actually an array before calling `map`

gkallenberg commented on routes/v1/search.js:
```diff
@@ -105,7 +104,7 @@ module.exports = function (app) {
     // display browse options and do other metrics related querying
     if ('aggregations' in params) {
-      params["aggregations"].map((agg) => {
+      params['aggregations'].map((agg) => {
```
Need to test whether `aggregations` is actually an array before calling `map`

gkallenberg commented on routes/v2/search.js:
```diff
@@ -0,0 +1,135 @@
+  // Filter block, used to search/filter data on specific terms. This can serve
+  // as the main search field (useful for browsing) but generally it narrows
+  // results in conjunction with a query
+  if ('filters' in params) {
```
Test for array before using `map`

gkallenberg commented on routes/v2/search.js:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  // Sort block, this orders the results. Can be asc/desc and on any field
+  if ('sort' in params) {
```
Test for array before using `map`

gkallenberg commented on routes/v2/search.js:
```diff
@@ -0,0 +1,135 @@
+  // but essentially this builds an object of record counts grouped by a term
+  // For example it can group works by authors/agents. This is used to
+  // display browse options and do other metrics related querying
+  if ('aggregations' in params) {
```
Test for array before using `map`

gkallenberg commented on routes/v2/v2.js:
```diff
@@ -0,0 +1,63 @@
+})
+
+const respond = (res, _resp, params) => {
+  let contentType = 'application/json'
```
use `const`

gkallenberg commented on routes/v1/work.js:
```diff
@@ -1,27 +1,26 @@
+    else if (respLen > 1) return handleError(res, new ElasticSearchError('Returned multiple records, identifier lacks specificity'))
 
-      let singleResp =  resp['hits']['hits'][0]['_source']
+    let singleResp = resp['hits']['hits'][0]['_source']
```
Use `const` instead of `let`

gkallenberg reviewed (changes requested):
Looking good. Just some minor adjustments I think make sense. Feel free to disagree or point out where I'm faulty in my assessment.

mwbenowitz committed:
SFR-291 Testing improvements

Fix a bug in v1/search introduced from merge with development and
make updates to package.json to fix bug that prevented coverage
report from being generated.

mwbenowitz committed:
SFR-291 Fix var/let assignment issues

mwbenowitz committed:
SFR-291 Fix const/let linting issues

mwbenowitz committed:
SFR-291 Minor fixes and array checking

Implement checking for arrays before parsing certain parameters
received in the search endpoints (for v1 and v2).

Additionally fix a bug with the swagger docs endpoint and remove
a console.log statement

mwbenowitz commented:
Thanks for the feedback, updated the endpoints to use const/let everywhere as it should be and linted the files to make sure that we aren't using `const` for anything that can get reassigned.

Also implemented `instanceof` checking before invoking `map` in both of the search endpoints to ensure that this executes safely.

gkallenberg commented on lib/client.js:
```diff
@@ -1,34 +1,34 @@
+    client.search({
+      q: 'gutenberg'
+    }).then(function (body) {
+      var hits = body.hits.hits
```
if hits doesn't get re-assigned, this can change to `const`.

gkallenberg reviewed (approved):
A minor adjustment to a `var` as well as changing .nvmrc and .travis.yml to point to the base Node version of 10.15.1; otherwise, good to go.

mwbenowitz committed:
Fix deployment node version and minor var change

Starting the express server would fail on node versions <= `6.10`.
Updating to node `10.15` or greater resolves the issue, so should
be set as such for CI as well as ebs deployment.

Finally, fix one lingering `var` that can be changed to `const`

mwbenowitz commented:
Merged in a final commit to updated node version to `10.15` and fix the lingering `var` issue that Greg identified
